### ISSUE TITLE
Observable.map iterators behave differently from regular Map

### ISSUE THREAD

**Issue by alexhisen** [mobxjs/mobx#2112]
At least in MobX 4.4.0, Observable.map iterators such as keys(), entries() and forEach() do not behave as those of a regular Map.

If map entries are removed during the iteration, a regular Map will still iterate correctly over all items that were originally in the Map. An Observable.map, on the other hand, will skip over items presumably due to internal index adjustments.

i.e. this works with a regular Map:
```
m = new Map();
m.set('key1', 'value1');
m.set('key2', 'value2');
for(key of m.keys()) { m.delete(key); };
// m.size() will now be 0
```
With an observable.map, the above code will leave an entry in the map.

A workaround is to first convert the map to an array with Array.from() and iterate over that.

At the least, this unexpected behavior should be documented on https://mobx.js.org/refguide/map.html

**Comment by spamforhope** [comment]
Hey!

I also found a bug with the Iterator Object.
If we do
```javascript
const itertatorTest = someObservableMap.values();

for (const item of itertatorTest) {
  // could be even empty
}

for (const item of itertatorTest) {
  // this loop won't work, ever
  // it thinks that itertatorTest hasn't finished and done() method didn't fire
  // so we'll never get next() method again
}
```

**Comment by FredyC** [comment]
That's not a bug, that's how iterators work. You cannot reuse the iterator. If you loop over it once, it will be marked as done. You have to call `.values()` again to get a fresh one.

Also, can you please set up a small reproduction (eg. with CodeSandbox) for the original issue?  Some simple comparisons of Map and observable wrapped Map. It will surely expedite the resolution.

**Comment by urugator** [comment]
Once #2287 is approved merged, I plan to optimize the map iterators and I think it will also solve this issue, I just don't want to include it into the mentioned PR as it's already quite big.

**Comment by urugator** [comment]
I've added following test:
```javascript
test("#2112 - iterators should be resilient to concurrent delete operation", () => {
    const map = mobx.observable.map([[1, 1], [2, 2], [3, 3]])
    const expectedMap = mobx.observable.map(map)
    for (const key of map.keys()) {
        map.delete(key)
        expectedMap.delete(key)
        expect(map.size).toBe(expectedMap.size)
        expect(Array.from(map)).toEqual(Array.from(expectedMap))
    }
})
```
And it's already passing on current master - I think it was fixed by #2279

**Comment by mweststrate** [comment]
Closing as fixed (afaik)
